src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -95,73 +97,143 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
-      user: this.user,
+    const studentsEnrollRequest: StudentsEnrollRequest = {
+      studentEnrollRequests: [],
     };
-    this.httpRequestService.post('/course/enrollSave', paramMap, this.enrollData)
-        .subscribe((resp: EnrollResultPanelList) => {
-          this.showEnrollResults = true;
-          this.statusMessage.pop(); // removes any existing error status message
-          this.statusMessageService.showSuccessMessage('Enrollment successful. Summary given below.');
-          this.enrollResultPanelList = resp.enrollResultPanelList;
-        }, (resp: ErrorMessageOutput) => {
-          this.statusMessage.pop(); // removes any existing error status message
-          this.statusMessage.push({
-            message: resp.error.message,
-            color: 'danger',
-          });
+
+    // Parse the user input to be requests.
+    for (const row of spreadsheetData) {
+      // Handsontable contains null value initially,
+      // see https://github.com/handsontable/handsontable/issues/3927
+      row.forEach((value: string, index: number) => {
+        if (value === null) {
+          row[index] = '';
+        }
+      });
+      const request: StudentEnrollRequest = {
+        section: row[hotInstanceColHeaders.indexOf(this.colHeaders[0])],
+        team: row[hotInstanceColHeaders.indexOf(this.colHeaders[1])],
+        name: row[hotInstanceColHeaders.indexOf(this.colHeaders[2])],
+        email: row[hotInstanceColHeaders.indexOf(this.colHeaders[3])],
+        comments: row[hotInstanceColHeaders.indexOf(this.colHeaders[4])],
+      };
+      studentsEnrollRequest.studentEnrollRequests.push(request);
+    }
+
+    this.studentService.enrollStudents(this.courseid, studentsEnrollRequest).subscribe((resp: Students) => {
+      const enrolledStudents: Student[] = resp.students;
+      this.showEnrollResults = true;
+      this.statusMessage.pop(); // removes any existing error status message
+      this.statusMessageService.showSuccessMessage('Enrollment successful. Summary given below.');
+      this.enrollResultPanelList =
+          this.populateEnrollResultPanelList(this.existingStudents, enrolledStudents,
+              studentsEnrollRequest.studentEnrollRequests);
+    }, (resp: ErrorMessageOutput) => {
+      this.statusMessage.pop(); // removes any existing error status message
+      this.statusMessageService.showErrorMessage(resp.error.message);
+    });
+    this.studentService.getStudentsFromCourse(this.courseid).subscribe((resp: Students) => {
+      this.existingStudents = resp.students;
+    });
+  }
+
+  private populateEnrollResultPanelList(existingStudents: Student[], enrolledStudents: Student[],
+                                         enrollRequests: StudentEnrollRequest[]): EnrollResultPanel[] {
+
+    const panels: EnrollResultPanel[] = [];
+    const studentLists: Student[][] = [];
+
+    for (const _ of Object.values(EnrollStatus).filter((value: EnrollStatus) => typeof value === 'string')) {
```

Comment:
What's the purpose of this loop if we are always pushing 5 empty arrays (corresponds to the number of enroll types in `enroll-status.ts`?

Reply:
Yes, the purpose here is just to push 5 empty arrays. Otherwise when I access something like `studentLists[3]` (which is `studentLists[EnrollStatus.ERROR]` in the code), it will throw error